### PR TITLE
upgrade certmanger k8s version in CI

### LINKED ISSUE #640: Failed to change internal tls from true to false
```
2021-05-11T10:52:23.783Z	ERROR	controller	Reconciler error	{"reconcilerGroup": "goharbor.io", "reconcilerKind": "Core", "controller": "core", "name": "harborcluster-sample-harbor-harbor", "namespace": "cluster-sample-ns", "error": "cannot set status to error: cannot set conditions to error: apply apps/v1, Kind=Deployment (cluster-sample-ns/harborcluster-sample-harbor-harbor-core): apply: Deployment.apps \"harborcluster-sample-harbor-harbor-core\" is invalid: [spec.template.spec.volumes[5].secret: Forbidden: may not specify more than 1 volume type, spec.template.spec.containers[0].volumeMounts[5].name: Not found: \"internal-certificates\"]: Deployment.apps \"harborcluster-sample-harbor-harbor-core\" is invalid: [spec.template.spec.volumes[5].secret: Forbidden: may not specify more than 1 volume type, spec.template.spec.containers[0].volumeMounts[5].name: Not found: \"internal-certificates\"]: apply apps/v1, Kind=Deployment (cluster-sample-ns/harborcluster-sample-harbor-harbor-core): apply: Deployment.apps \"harborcluster-sample-harbor-harbor-core\" is invalid: [spec.template.spec.volumes[5].secret: Forbidden: may not specify more than 1 volume type, spec.template.spec.containers[0].volumeMounts[5].name: Not found: \"internal-certificates\"]: Deployment.apps \"harborcluster-sample-harbor-harbor-core\" is invalid: [spec.template.spec.volumes[5].secret: Forbidden: may not specify more than 1 volume type, spec.template.spec.containers[0].volumeMounts[5].name: Not found: \"internal-certificates\"]", "errorVerbose": "Deployment.apps \"harborcluster-sample-harbor-harbor-core\" is invalid: [spec.template.spec.volumes[5].secret: Forbidden: may not specify more than 1 volume type, spec.template.spec.containers[0].volumeMounts[5].name: Not found: \"internal-certificates\"]: Deployment.apps \"harborcluster-sample-harbor-harbor-core\" is invalid: [spec.template.spec.volumes[5].secret: Forbidden: may not specify more than 1 volume type, spec.template.spec.containers[0].volumeMounts[5].name: Not found: \"internal-certificates\"]\napply\ngithub.com/goharbor/harbor-operator/pkg/controller.(*Controller).applyAndCheck\n\t/home/runner/work/harbor-operator/harbor-operator/pkg/controller/common.go:167\ngithub.com/goharbor/harbor-operator/pkg/controller.(*Controller).ProcessFunc.func1\n\t/home/runner/work/harbor-operator/harbor-operator/pkg/controller/resource.go:145\ngithub.com/goharbor/harbor-operator/pkg/graph.(*resourceManager).Run.func1\n\t/home/runner/work/harbor-operator/harbor-operator/pkg/graph/runner.go:42\ngolang.org/x/sync/errgroup.(*Group).Go.func1\n\t/home/runner/go/pkg/mod/golang.org/x/sync@v0.0.0-20201020160332-67f06af15bc9/errgroup/errgroup.go:57\nruntime.goexit\n\t/opt/hostedtoolcache/go/1.15.11/x64/src/runtime/asm_amd64.s:1374\napply apps/v1, Kind=Deployment (cluster-sample-ns/harborcluster-sample-harbor-harbor-core)\ngithub.com/goharbor/harbor-operator/pkg/controller.(*Controller).ProcessFunc.func1\n\t/home/runner/work/harbor-operator/harbor-operator/pkg/controller/resource.go:147\ngithub.com/goharbor/harbor-operator/pkg/graph.(*resourceManager).Run.func1\n\t/home/runner/work/harbor-operator/harbor-operator/pkg/graph/runner.go:42\ngolang.org/x/sync/errgroup.(*Group).Go.func1\n\t/home/runner/go/pkg/mod/golang.org/x/sync@v0.0.0-20201020160332-67f06af15bc9/errgroup/errgroup.go:57\nruntime.goexit\n\t/opt/hostedtoolcache/go/1.15.11/x64/src/runtime/asm_amd64.s:1374\ncannot set status to error: cannot set conditions to error: apply apps/v1, Kind=Deployment (cluster-sample-ns/harborcluster-sample-harbor-harbor-core): apply: Deployment.apps \"harborcluster-sample-harbor-harbor-core\" is invalid: [spec.template.spec.volumes[5].secret: Forbidden: may not specify more than 1 volume type, spec.template.spec.containers[0].volumeMounts[5].name: Not found: \"internal-certificates\"]: Deployment.apps \"harborcluster-sample-harbor-harbor-core\" is invalid: [spec.template.spec.volumes[5].secret: Forbidden: may not specify more than 1 volume type, spec.template.spec.containers[0].volumeMounts[5].name: Not found: \"internal-certificates\"]\ngithub.com/goharbor/harbor-operator/pkg/controller.(*Controller).HandleError\n\t/home/runner/work/harbor-operator/harbor-operator/pkg/controller/errors.go:64\ngithub.com/goharbor/harbor-operator/pkg/controller.(*Controller).Reconcile\n\t/home/runner/work/harbor-operator/harbor-operator/pkg/controller/common.go:150\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:235\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:209\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:188\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1\n\t/home/runner/go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:155\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil\n\t/home/runner/go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:156\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/home/runner/go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:133\nk8s.io/apimachinery/pkg/util/wait.Until\n\t/home/runner/go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:90\nruntime.goexit\n\t/opt/hostedtoolcache/go/1.15.11/x64/src/runtime/asm_amd64.s:1374"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:209
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker
	/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:188
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1
	/home/runner/go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:155
k8s.io/apimachinery/pkg/util/wait.BackoffUntil
	/home/runner/go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:156
k8s.io/apimachinery/pkg/util/wait.JitterUntil
	/home/runner/go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:133
k8s.io/apimachinery/pkg/util/wait.Until
	/home/runner/go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:90
W0511 10:53:03.947141       1 warnings.go:70] networking.k8s.io/v1beta1 Ingress is deprecated in v1.19+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
```

### LINKED ISSUE #649: Adjust the verifying Kubernetes versions to 1.19, 1.20 and 1.21
1.18 will be out of the verification list (also in CI pipeline)

### PR DESCRIPTION
**Pull request by bitsf**
close #649 
close #640
1. upgrade cert manager to 1.3.1
2. upgrade kind to 0.11.0
3. upgrade k8s to 1.21.1
4. upgrade ingress to 0.46.0

### CODE REVIEW

**Comment by bitsf** [comment]
k8s 1.21.1 failed
image

File: pkg/controller/context.go [hunk 1]
```diff
@@ -9,8 +9,7 @@ import (
 	ctrl "sigs.k8s.io/controller-runtime"
 )
 
-func (c *Controller) NewContext(req ctrl.Request) context.Context {
-	ctx := context.TODO()
+func (c *Controller) NewContext(ctx context.Context, req ctrl.Request) context.Context {
```
Comment (holyhope) [inline pkg/controller/context.go:12]: IMO `NewContext` expecting a context as first parameter is odd.
I suggest to rename it.
```suggestion
func (c *Controller) PopulateContext(ctx context.Context, req ctrl.Request) context.Context {
```

File: pkg/controller/ready.go [hunk 2]
```diff
@@ -20,17 +21,14 @@ func (c *Controller) ensureResourceReady(ctx context.Context, res *Resource) err
...
-	if err != nil {
-		return serrors.UnrecoverrableError(err, serrors.OperatorReason, "cannot get object key")
-	}
+	objectKey := client.ObjectKeyFromObject(res.resource)
 
 	gvk := c.AddGVKToSpan(ctx, span, res.resource)
 	l := logger.Get(ctx)
 
-	result := res.resource.DeepCopyObject()
+	result := res.resource.DeepCopyObject().(resources.Resource)
```
Comment (holyhope) [inline pkg/controller/ready.go:29]: Why do we need to cast it?
Does the following not work?
```suggestion
	var result resources.resource = res.resource.DeepCopyObject()
```

File: pkg/controller/secret-format.go [hunk 3]
```diff
@@ -51,7 +47,6 @@ func (c *Controller) EnsureSecretType(ctx context.Context, node graph.Resource)
 	case res.resource.(*corev1.Secret).Type:
 		return nil
 	case corev1.SecretTypeOpaque:
-		// nolint:exhaustive
```
Comment (holyhope) [inline pkg/controller/secret-format.go:54]: Did it become exhaustive?

**Review by holyhope — APPROVED** [review]
Great work. So much small but satisfying improvements!

File: pkg/controller/context.go (hunk 1 above)
Comment (bitsf) [reply pkg/controller/context.go:12]: ok

File: pkg/controller/secret-format.go (hunk 3 above)
Comment (bitsf) [reply pkg/controller/secret-format.go]: it seems outer nolint is enough, anyway I add it back

File: pkg/controller/ready.go (hunk 2 above)
Comment (bitsf) [reply pkg/controller/ready.go:29]: because DeepCopyObject() is returning running.Object not client.Object

Comment (bitsf) [reply pkg/controller/ready.go:29]: your suggestion code is not working

**Review by steven-zou — APPROVED** [review]
Great job! @bitsf

**Pull request merged**